UCOSP project - File coverage view

LinkaiQi opened the pull request:
### File coverage
Show coverage for file at a given revision.


#### In order to show coverage information, file-coverage-view needs data from both hg and ActiveData:
- Source code from hg. (https://hg.mozilla.org/integration/mozilla-inbound/raw-file/...)
- Coverage data from ActiveData. (https://activedata.allizom.org)


#### Features:
1. File-coverage-view highlights covered and uncovered line of a file.
2. Show coverage percentage of the file with dynamic color.
3. Show a number in each line represents the number of test-suite that cover the line.
4. By default it shows all tests that cover the file in the RHS ```covered-tests-sidebar```. If a line has been selected, the file-coverage-view will display which suite/chunk that covers the selected line with info of ```suite-name```, ```chunk-number``` and ```platform```.

#### PDF doc demo:
https://github.com/klahnakoski/firefox-code-coverage-frontend/blob/ucosp/docs/File_Coverage_View.pdf

#### Contributor:
**Mentor**: Kyle Lahnakoski (klahnakoski@mozilla.com)
Jennifer Yuen  (jyuen1@ualberta.ca)
Linkai Qi  (linkai@ualberta.ca)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/firefox-code-coverage-frontend/76)
<!-- Reviewable:end -->

klahnakoski commented on src/utils/fetch_data.js:
```diff
@@ -8,6 +9,39 @@ const jsonHeaders = {
   Accept: 'application/json',
 };
 
+async function httpFetch(params) {
```
You want this function inlined? That seems extreme since it does a number of mundane checks.  Also, how is it not a pure function?

armenzg replied:
From reviewing the definition even making an HTTP request disqualifies it.
I was thinking specifically of logging errors.

Let me seat on this.

yuenj commented:
Hi Kyle, Armen, 
Just made a commit to address most of the feedback. However, I wasn't sure how to merge fileviewer and diffviewer because each retrieves from a different data source:

> Please take the code from
https://github.com/mozilla/firefox-code-coverage-frontend/blob/master/src/components/diffviewer.js#L31
and move into data.js
Make your code and the code from diffviewer use the same code path.

Please let me know your thoughts.  

(p.s. Linkai and I are alternating addressing the feedback)

klahnakoski commented on src/components/fileviewer.js:
```diff
@@ -20,157 +16,80 @@ export default class FileViewerContainer extends Component {
       /* app status */
       status: {
         app: undefined,
-        fetch: {
```
I believe you must keep `fetch`, so that you can set it later.

marco-c commented:
> Just made a commit to address most of the feedback. However, I wasn't sure how to merge fileviewer and diffviewer because each retrieves from a different data source:

I think we should make this code retrieve the data from the shipit backend, like the diffviewer does. The shipit backend can then either use codecov or activedata.

armenzg commented on src/components/fileviewer.js:
```diff
@@ -0,0 +1,194 @@
+const FileViewerMeta = ({ revision, path, status, parsedFile, coverage }) => {
+  const showStatus = (label, fetched) => {
+    let msg;
+    if (fetched === undefined) {
```
Fetched is always either true or false but never undefined.
Please remove this case.

klahnakoski replied:
`fetched`, at acb9730df71 (before the refactor) was originally set to `undefined` before fetching of coverage and source code.  For some reason `parsedFile.length > 0` is being used; which makes the assumption that the file has lines, but also shows an error until the source code arrives.  `Object.keys(coverage).length` assumes there is going to be coverage, and also shows an error until coverage arrives. 

The original trinary logic should be kept.

armenzg commented on src/components/fileviewer.js:
```diff
@@ -0,0 +1,194 @@
+      <div>
+        <div className="file-view">
+          <FileViewerMeta
+            revision={revision}
```
Another easier way to pass state down and letting the component grab what it wants is by doing this:
```javascript
<FileViewerMeta
   ...this.state
/>
```

armenzg commented on src/components/fileviewer.js:
```diff
@@ -0,0 +1,194 @@
+      /* app status */
+      status: {
+        app: undefined,
+      },
```
Could you please flatten this to `appError: ''`? (instead of two levels)

armenzg commented on src/components/fileviewer.js:
```diff
@@ -0,0 +1,194 @@
+        </div>
+        <div className="file-meta-title">File Coverage</div>
+      </div>
+      {status.app && <span className="error_message">{status.app}</span>}
```
We're not displaying an error I'm hitting from HG.
When I pass a bad revision I get this message within the FileViewer:
> error: revision not found: f7

http://localhost:3001/file?revision=f7&path=dom/events/EventStateManager.cpp
https://hg.mozilla.org/integration/mozilla-inbound/raw-file/f7/dom/events/EventStateManager.cpp

If I put a bad path I get this instead:
> error: dom/events/EventStateManager.cp@f763151e6c59: not found in manifest

http://localhost:3001/file?revision=f763151e6c59fd0e18c69a918a36d2e6741d5b0e&path=dom/events/EventStateManager.cp

klahnakoski replied:
Revision acb9730 (before the refactor) was showing an error if revision was not found.  It also showed an error when the path was not found.

klahnakoski replied:
... both showed the same:

![image](https://user-images.githubusercontent.com/2334429/33442655-acd1d268-d5c3-11e7-9812-5b5443c3ff61.png)

yuenj replied:
Should I restore status: { app: .., fetch: { source: .., coverage: .. } } for now until we find a better solution? Does fetch status of source and coverage belong to the UI and can be represented in state?

klahnakoski replied:
In general the "correct" place for a state variable is in the most specific component that holds all usages of that state.  This may be why we had problems with `setState()` on deep structures; the state belongs all the way down in `<li className="file-meta-li">` component where it would be defined as a top-level (not nested) property.  

Maybe the `<li>` is a bit much, another level up at `FileViewerMeta` where there is an actual component you can set state on.

armenzg commented on src/components/fileviewercov.js:
```diff
@@ -0,0 +1,119 @@
+    );
+  }
+  return (
+    <div className="tests_viewer">
```
Change all className invocations to have classes that use "-" instead of "_".

armenzg commented on src/components/fileviewercov.js:
```diff
@@ -0,0 +1,119 @@
+
+  if (!coverage) {
+    content = <h3>Fetching coverage from backend...</h3>;
+  } else if (!lineNumber) {
```
Please don't have 4 different sections intermixing the DOM with variables and wrapping it again on the return at the bottom of the function.

Calculate the content of the h3 tag and if there should be any ul content.

You can do this instead:
```javascript
return (
  <div className="tests-viewer-title">Covered Tests</div>
     <h3>{h3Text}</h3>
     {ulText}
  </div>
);
```

armenzg commented on src/components/fileviewercov.js:
```diff
@@ -0,0 +1,119 @@
+class Test extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { expand: false };
```
Make `expand` be a prop and let the calling component determine what to send down.

armenzg commented on src/components/fileviewercov.js:
```diff
@@ -0,0 +1,119 @@
+    this.handleTestOnClick = this.handleTestOnClick.bind(this);
+  }
+
+  handleTestOnClick() {
```
This should be handled on the parent container.
Pass the function as a prop.

klahnakoski replied:
I do not understand: `handleTestOnClick()` only changes this' component state.  Why would the parent container get involved?

armenzg replied:
You're right.
The pattern I had in mind is when state is shared among childen. In here is completely isolated.

These are the concepts I mixed in head:
https://reactjs.org/docs/lifting-state-up.html
https://reactjs.org/docs/forms.html

armenzg commented on src/components/fileviewercov.js:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  render() {
+    const test = this.props.test;
```
Use JS's object destructuring.
```javascript
const { test } = this.props;
```

armenzg commented on src/components/fileviewercov.js:
```diff
@@ -0,0 +1,119 @@
+
+/* shows coverage percentage of a file */
+export const CoveragePercentageViewer = ({ coverage }) => {
+  let percentageCovered;
```
`const` unless I'm reading the code incorrectly.

armenzg commented on src/components/fileviewercov.js:
```diff
@@ -0,0 +1,119 @@
+  return (
+    <div className="coverage_meta">
+      <div className="coverage_meta_totals">
+        {percentageCovered &&
```
I would move this conditional up and wrap everything inside of it.

armenzg commented on src/components/fileviewercov.js:
```diff
@@ -0,0 +1,119 @@
+
+  return (
+    <div className="coverage_meta">
+      <div className="coverage_meta_totals">
```
No need for two divs.

armenzg commented on src/style.css:
```diff
@@ -149,3 +156,184 @@ pre { margin: 0; }
+	display: block;
+	margin-top: 10px;
+}
+tr.file_line {
```
I think you can do:
```css
.file_line {
  :hover {.....}
  :selected {....}
}
```
Don't quote me though.

armenzg commented on src/utils/data.js:
```diff
@@ -36,6 +37,37 @@ const coverageSummary = (coverage) => {
+      if (!s.testsPerHitLine[line]) {
+        s.uncoveredLines.push(line);
+      }
+    });
```
Why are the last 5 lines within the 1st forEach loop?

yuenj replied:
It's due to the nature of the object returned by the query to Active Data.  If I'm correct, the outer loop iterates over each test metadata, and the inner loop iterates over the uncovered lines that that test reports.

armenzg replied:
I think this would work:
```javascript
coverage.forEach((c) => {
  c.source.file.covered.forEach((line) => {
    s.coveredLines.push(line);
    if (!s.testsPerHitLine[line]) {
      s.testsPerHitLine[line] = [];
    }
    s.testsPerHitLine[line].push(c);
  };
  // get uncovered lines
  c.source.file.uncovered.forEach((line) => {
    if (!s.testsPerHitLine[line]) {
      s.uncoveredLines.push(line);
    }
  });
});
```

armenzg commented on src/utils/data.js:
```diff
@@ -109,3 +141,39 @@ export const csetWithCcovData = async (cset) => {
+      where: {
+        and: [
+          { eq: { 'source.file.name': `${path}` } },
+          { eq: { 'repo.changeset.id12': `${revision}` } },
```
How do we specify the repoName?
I'm surprised we're using 'mozilla-inbound' as the reference when all coverage data is supposed to be coming from 'mozilla-central'.

klahnakoski replied:
@LinkaiQi, @yuenj  you can add `{eq:{'repo.branch.name': `${repoName}`}}` to the where clause

armenzg commented on src/utils/fetch_data.js:
```diff
@@ -8,14 +9,25 @@ const jsonHeaders = {
   fetch(`${ccovBackend}/coverage/changeset_summary/${changeset}`, { jsonHeaders });
+
+// query active data
+export const query = body =>
```
Please name this `queryActiveData`.

armenzg commented on src/utils/data.js:
```diff
@@ -109,3 +141,39 @@ export const csetWithCcovData = async (cset) => {
+  }
+};
+
+export const fileRevisionWithActiveData = async (revision, path) => {
```
Name this `queryFileCoverage`

armenzg commented on src/utils/fetch_data.js:
```diff
@@ -8,14 +9,25 @@ const jsonHeaders = {
 
-export const getDiff = changeset =>
-  fetch(`${hgHost}/mozilla-central/raw-rev/${changeset}`, { plainHeaders });
+const jsonPost = params =>
```
`({ url, body })`

armenzg commented on src/utils/fetch_data.js:
```diff
@@ -8,14 +9,25 @@ const jsonHeaders = {
+export const getDiff = (changeset, repoPath) =>
+  fetch(`${hgHost}/${repoPath}/raw-rev/${changeset}`, { plainHeaders });
+
+// raw-file fetcher (fileviewer)
```
No need for a comment.

armenzg commented on src/utils/fetch_data.js:
```diff
@@ -8,14 +9,25 @@ const jsonHeaders = {
 export const getChangesetCoverageSummary = changeset =>
   fetch(`${ccovBackend}/coverage/changeset_summary/${changeset}`, { jsonHeaders });
+
+// query active data
```
No need for a comment.

yuenj commented on src/components/fileviewer.js:
```diff
@@ -0,0 +1,194 @@
+        testsPerHitLine: [],
+      },
+      selectedLine: undefined,
+    };
```
We had a discussion about this and the reason was for speed.  We thought that pre-constructing the hash map and arrays for "testsPerHitLine", "coveredLines", etc. would allow faster retrieval of this information later, rather than recomputing this information each time the DOM was re-rendered.   Do you have any suggestions on where we can store this global information outside of state in React?

armenzg replied:
If you use `...this.state` when declaring a component (as mentioned in [here](https://github.com/mozilla/firefox-code-coverage-frontend/pull/76/files#r153851836)) you won't need to declare anything in here with this.state.

yuenj replied:
Yes but say the data returned from queryActiveData() is some large object.  Say clicking on a line causes a re-rendering because this.state.selectedLine is updated and we need selectedLine to be in the state because it's used between components (though probably a better design is to trigger a method).  This causes the DOM to re-render so items such as the coverage percentage will have to be recalculated from the large query object vs. caching it some place.

(I'm assuming ...this.state just passes the entire state down to the components instead of individually)

armenzg replied:
It actually uses JS deconstructing.
It is the same as
```javascript
this.state = {
  foo: 'bar',
  xyz: 'abc',
}
.......
<FooBar
  foo={this.state.foo}
  xyz={this.state.xyz}
/>
// equivalent to
<FooBar
  {...this.state}
/>
```

Now the component might not need all properties of the state but it can pick the ones it wants (e.g. { xyz }) in its declaration.

For a moment, you can forget about the spreader syntax (...).
If you left the render() method as is, you still don't need to define `this.state` within the constructor since you're doing this inside of render:
> const { status, revision, path, parsedFile, coverage, selectedLine } = this.state;

It would be like passing underfined to all of them.

Now, in some cases you initialize to `[]` instead of `undefined`. Passing undefined to subcomponents that might try `.length > 0` could be problematic.
However, in those cases you can do this inside of the main render():
```javascript
{(parsedFile) && /* All components that need parsedFile to be ready */}
```

yuenj commented on src/components/fileviewer.js:
```diff
@@ -0,0 +1,194 @@
+    this.setState({ selectedLine: selectedLineNumber });
+  }
+
+  parseQueryParams() {
```
Currently parseQueryParams() is called from ComponentWillMount().  The documentations say that if it is called synchronously then this should not cause an extra re-rendering.  However, we'll move it into this.state directly.

armenzg replied:
Do you have a reference to that?

Calling this.setState() will cause an extra re-rendering if there's any state changes.

yuenj replied:
https://reactjs.org/docs/react-component.html

componentWillMount() is invoked immediately before mounting occurs. It is called before render(), therefore calling setState() synchronously in this method will not trigger an extra rendering. Generally, we recommend using the constructor() instead.

armenzg replied:
Thank you! I was not aware of that detail. In any case I was thinking of component*Did*Mount().
As they recommend, let's use constructor() unless there's a technichal reason preventing us to.

yuenj replied:
ok

yuenj commented:
> @yuenj this is a big refactor: Big refactors are difficult to review, and prevent others from making changes until they are merged. I suggest re-doing this as multiple commits. One where functions are moved (but not changed), another that changes the method signatures (but has little functional effect), and a final one that makes improvements.

@klahnakoski do you still want me to do this since it seems Armen has reviewed it?

yuenj commented:
> Could you please give me revision and path that actually show me a test associated to it?
> I was expecting this URL to work:
> http://localhost:3001/file?revision=f763151e6c59fd0e18c69a918a36d2e6741d5b0e&path=dom/events/EventStateManager.cpp

http://localhost:3000/file?revision=cff54d05bcad&path=accessible/atk/Platform.cpp

> If a revision/path does not have any tests covering it, is there worth on showing the UI at all?

Should it be up to the user to decide that?  Or we can disable entry to it from diffviewer or other entry points?

armenzg commented:
> > Could you please give me revision and path that actually show me a test associated to it?
> > I was expecting this URL to work:
> > http://localhost:3001/file?revision=f763151e6c59fd0e18c69a918a36d2e6741d5b0e&path=dom/events/EventStateManager.cpp
> 
> http://localhost:3000/file?revision=cff54d05bcad&path=accessible/atk/Platform.cpp
> 
> > If a revision/path does not have any tests covering it, is there worth on showing the UI at all?
> 
> Should it be up to the user to decide that?  Or we can disable entry to it from diffviewer or other entry points?  
For now, let's make it clear so they don't waster their time trying to clicking on all lines in hope to find something useful.

armenzg commented:
I've requested Kyle's repo to be updated to the latest code:
https://github.com/klahnakoski/firefox-code-coverage-frontend/pull/54

Once that is merged could you please `git rebase master` for the `ucosp` branch?
This might be a big task as I assume there will be lost of conflicts.

This will solve the message on this PR that says "This branch cannot be rebased due to conflicts"

klahnakoski commented:
> @klahnakoski do you still want me to do this since it seems Armen has reviewed it?

@yuenj if armenzg can follow along with the changes you made, then o not revert it, keep going forward.

armenzg commented on src/components/fileviewercov.js:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  handleTestOnClick() {
+    this.expandClass = (this.state.expand) ? '' : 'expanded';
```
Store this information as state rather than an object's property.
The UI is the representation of the state.

klahnakoski commented on src/components/fileviewer.js:
```diff
@@ -0,0 +1,194 @@
+
+  async componentDidMount() {
+    /* Get source code and coverage in parallel  */
+    await Promise.all([this.fetchSourceCode(), this.fetchCoverage()]);
```
> Also, if you think about it, within different a Promise.all() there are calls to this.setState() from each of the two functions which is a bad practice since it causes race conditions and could cause multiple re-renderings of the app.

I do not think this is correct. Javascript is single threaded, so there are no race conditions, React is great because it does not re-render the whole app; it only makes the necessary changes.

armenzg replied:
Yes, that's correct. Thanks for correcting it.

marco-c replied:
JavaScript is single threaded, but the order in which the fetches finish is not defined. If you have fetchA using setState({ obj: A }) and fetchB using setState({ obj: B }), the final state could be A or B according to which fetch finishes first.
In this case, there's a race condition when both fetches fail (so you can't be sure if the status is going to be `We did not manage to fetch source file from hg.mozilla` or `We did not manage to fetch test coverage from ActiveData`).

klahnakoski replied:
That race condition happens when both fail. No matter which error shows, it is still not misleading.  the log should have a full accounting of what happened, even if the order is not defined.

marco-c replied:
Yes, the consequences of the race condition are not too bad, but I would suggest avoiding it anyway. You never know how the code is going to evolve.

LinkaiQi replied:
"Promise.all waits for all fulfillments (or the first rejection)." 

If we do this:
```js
const [result1, result2] = await Promise.all(request1, request2)
if (success) {
    this.setState(
        result1,
        result2,
    );
}
```
Correct me if I'm wrong, the ```this.setState()``` will be executed only if the two network requests(p1, p2) get response, which contrary the reason we use the ```promise.all()```.
The reason we use the ```promise.all()```method is to allow making network requests in parallel and when one of them are done, they don't need to wait for the other one to call ```this.setState()```.

marco-c replied:
You simply need a try catch around `await Promise.all` to check for the error.

armenzg replied:
You got it right.
I specifically want us to wait for both to complete before proceeding.
From what I remember, the UI is not useful without both pieces of data fetched.
Specifically the raw file.

Besides, the code is also easier to read with this approach. The logic to fetch data is not mixed with the logic to evaluate data and setting the state. Separation of responsibilities.

armenzg commented on src/index.js:
```diff
@@ -1,10 +1,10 @@
 
 ReactDOM.render(
-  <HashRouter><App /></HashRouter>,
+  <BrowserRouter><App /></BrowserRouter>,
```
Is this change necessary?
I needed to change to HashRouter because it worked much better.

You can use this with HashRouter:
`const urlParams = new URLSearchParams(props.location.search);`

klahnakoski replied:
@LinkaiQi please change this to use the Hashrouter.  It was my mistake to ask for "?" parameters instead of using "#" fragments

armenzg replied:
"?" parameters work for both.
> #/file?revision=foo&path=bar
> /file?revision=foo&path=bar

I found this:
> BrowserRouter should be used when you have a server that will handle dynamic requests (knows how to respond to any possible URI), while HashRouter should be used for static websites (can only respond to requests for files that it knows about).
> 
> Usually it is preferable to use a BrowserRouter, but if your website will be hosted on a server that only serves static files, then HashRouter is a good solution.

These are the notes as to why I needed it:
https://github.com/armenzg/firefox-code-coverage-frontend/commit/0107a1fc3a49a080f9aa3b370a4e24bcfde6ac65

klahnakoski replied:
> "?" parameters work for both.
>
> >   #/file?revision=foo&path=bar
> >    /file?revision=foo&path=bar

Well, those two `?` are not the same; the former is in the URL fragment, which is what I wanted to avoid.  `?` in the fragment is very confusing, at least for me: It took me over an hour, and questions to @armenzg, to figure out how to compose a proper URL to use the file viewer because of the reversal of `?` and `#`.

LinkaiQi commented:
Hi, Armen, Kyle, Marco

Code has been updated, hopes it addresses all the issues/feedbacks mentioned above.
Please let me and Jen know if anything can be improved.

Here is a URL you can test it: 
http://localhost:3000/#/file?revision=b4cef8d1dff0&path=/accessible/atk/Platform.cpp

Best,
Linkai Qi

klahnakoski commented:
I humbly suggest next time we have a PR this big; the review comments be turned into issues for the source branch to resolve, leaving the PR with much less to talk about.

marco-c commented:
> I humbly suggest next time we have a PR this big; the review comments be turned into issues for the source branch to resolve, leaving the PR with much less to talk about.

Or, avoid too large PRs and land individual and separate changes.

klahnakoski commented:
> > I humbly suggest next time we have a PR this big; the review comments be turned into issues for the source branch to resolve, leaving the PR with much less to talk about.

> Or, avoid too large PRs and land individual and separate changes.

Yes, that too. 

At the beginning of the term, we thought keeping the issues and churn on a separate fork would make tracking development of the two different views (coverage by changeset and coverage by file) easier.  I still agree that this was a good choice.  I just wish I redirected Armen's reviews to Issues right away.

klahnakoski commented on src/components/fileviewercov.js:
```diff
@@ -94,21 +93,22 @@ export const Test = ({ row, test, expand, handleTestOnExpand }) => {
-    <div className="coverage_meta">
-      <div className="coverage_meta_totals">
-        {percentageCovered &&
-          <span className="percentage_covered" style={{ backgroundColor: `${Color.getPercentCovColor(percentageCovered)}` }}>
```
I am sad to see the color go!  Oh well.

armenzg commented:
I will be using reviewable.io to help me distinguish what has changed since the last review.
I believe I have enabled the two way configuration. We will see.

Good job on addressing the issues.

This is very close. I think we might only need 1 or 2 more rounds.
If you have a percentage of issues addressed by the end of your night please push a working state of the branch and ask for feedback. I would rather tackle one batch at a time.

Please check the output of `yarn lint` and rebase this branch over `master`.

Also, please load the URL you mentioned and look at the browser's console. Fix all errors there.

Once I load the given URL I have 80 white lines. Any objections to have a follow up issue that hides those lines?

We should also file an issue to be able to navigate to this view from the DiffViewerContainer.

Kyle, would you do the honor of filing these two issues if you agree?

I must say I was wrong with regards to our conversation around race conditions.
I've tried the app and fetching code coverage takes long enough that it makes you feel that the app is not working.
If at least the raw file was shown we could make it feel that something is actually happening.
https://github.com/mozilla/firefox-code-coverage-frontend/issues/78

---

Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 26 unresolved discussions.

---

*[docs/diffview_tests_prototype.jpeg, line 0 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-gZFLtqNEfss99v-b5:-L-gZFLuX1EOdDoVPY8w:btdaiir) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/docs/diffview_tests_prototype.jpeg#L0)):*
Please remove the .jpeg and .pdf files if they're not include in .md files.
In fact, create a new PR about fixing it the right way.
Use this branch (instead of master) as a starting point for the new PR.
See http://docutils.sourceforge.net/docs/user/rst/quickref.html

---

*[src/style.css, line 17 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-gZywwJqiqHtaFhWYI:-L-gZywwJqiqHtaFhWYJ:b-807afv) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/style.css#L17)):*
> ```CSS
> 
> /*
>  * CSS for diff viewer and file virwer
> ```

Typo.

---

*[src/style.css, line 137 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-g_6uvYw-z_riuAdBG:-L-g_6uvYw-z_riuAdBH:baq01v2) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/style.css#L137)):*
> ```CSS
> 
> /* Changesets */
> .hidden_changeset {
> ```

Please keep using '-' instead of '_'. Consistency with the rest of the files.
My apologies if I said otherwise in any of my other comments.

I've been trying to follow https://udacity.github.io/frontend-nanodegree-styleguide/css.html#delimiters

---

*[src/style.css, line 185 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-g_Zb9MznKP76Tyox-:-L-g_ZbAAiNKN9ZqInWf:brc6l51) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/style.css#L185)):*
> ```CSS
> .file-view {
> 	display: inline-block;
> 	min-width: 1000px;
> ```

Are you sure we want to have a minimum width? That seems like a very high value.

---

*[src/style.css, line 209 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-g_hSGvoC7dBKiLbEa:-L-g_hSGvoC7dBKiLbEb:bouxsom) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/style.css#L209)):*
> ```CSS
> 	color: Grey;
> 	text-align: center;
> 	/*padding: 1px;*/
> ```

Remove commented out code.

---

*[src/style.css, line 216 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-g_n6vySOrdSuCxzzb:-L-g_n6vySOrdSuCxzzc:b-tx2s5z) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/style.css#L216)):*
> ```CSS
> span.tests {
> 	background: #21ba45;
>   border-radius: 10px;
> ```

Please keep consistent indentation.

---

*[src/style.css, line 263 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-g_vv8dYfAGiPjqYtK:-L-g_vv8dYfAGiPjqYtL:bwmg60m) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/style.css#L263)):*
> ```CSS
> }
> .toggleable-test-title {
> 	/*border-top: 1px solid #CCC;*/
> ```

Commented out style.

---

*[src/components/diffviewer.js, line 28 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-gb4rNzoSLrmfot0t_:-L-gb4rNzoSLrmfot0ta:b-n9rw6q) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/diffviewer.js#L28)):*
> ```JavaScript
>   async componentDidMount() {
>     const { changeset } = this.props;
>     await Promise.all([this.fetchCsetData(changeset), this.fetchDiff(changeset)]);
> ```

Thank you for fixing this,

---

*[src/components/diffviewer.js, line 31 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-gbDVv1Dld9YEGDV4A:-L-gbDVv1Dld9YEGDV4B:b-xx2vip) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/diffviewer.js#L31)):*
> ```JavaScript
>   }
> 
>   async fetchCsetData(changeset) {
> ```

Please rename this to fetchSetCoverageData()

---

*[src/components/diffviewer.js, line 42 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-gbalor375IRNgxyn2:-L-gbalor375IRNgxyn3:b-pws31p) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/diffviewer.js#L42)):*
> ```JavaScript
>   }
> 
>   async fetchDiff(changeset) {
> ```

Rename this to `fetchSetDiff`

---

*[src/components/diffviewer.js, line 44 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-gaPiG0PoCq0JqpbfM:-L-gaPiG0PoCq0JqpbfN:blc11z) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/diffviewer.js#L44)):*
> ```JavaScript
>   async fetchDiff(changeset) {
>     try {
>       const text = await (await FetchAPI.getDiff(changeset, 'mozilla-central')).text();
> ```

Do not use magic values.

Add to `getDiff()` (not this function) specifying the repoName as part of the signature of the function.
We can use a default value there.

---

*[src/components/fileviewer.js, line 22 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-gcV4TbSssyukSSQgg:-L-gcV4UwwdQoPaeKvsb:b-oiy6or) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewer.js#L22)):*
> ```JavaScript
>       parsedFile: undefined,
>       coverage: undefined,
>     };
> ```

I'm looking at the render method and I don't see you using anything like `parsedFile.length` or `Object.keys(parsedFile)`.
Therefore, there's no need to define anything in here as `undefined`.
In other words, do not add a line with `this.state` at all.
I believe I explained this in the past but I can't be sure if I made it clear.

---

*[src/components/fileviewer.js, line 64 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hQjo3zV5KPvCaoQtJ:-L-hQjo3zV5KPvCaoQtK:b-rbvbq2) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewer.js#L64)):*
> ```JavaScript
>     const parsedQuery = queryString.parse(this.props.location.search);
>     if (!parsedQuery.revision || !parsedQuery.path) {
>       this.state = { ...this.state, appErr: "Undefined URL query ('revision', 'path' fields are required)" };
> ```

Webpack is warning in the console about  modifying the state directly.

Please return data from this function and set the state only once in the constructor:
```
const { appError, revision, path } = this.parseQueryParams();
```
or
```
this.state = this.parseQueryParams();
```

---

*[src/components/fileviewer.js, line 102 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hS_he_c3XZKs-zATl:-L-hS_he_c3XZKs-zATm:bpl6lj0) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewer.js#L102)):*
> ```JavaScript
>         parsedFile.map((lineText, i) => (
>           <Line
>             key={lineText}
> ```

Use {i + 1} instead. It will remove all flattenChildren() errors from the console.

---

*[src/components/fileviewer.js, line 155 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hCIGseFZi6vW7A_tq:-L-hCIGt7Hdy58dZ1dG7:b3twxqc) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewer.js#L155)):*
> ```JavaScript
>     {appErr && <span className="error-message">{appErr}</span>}
> 
>     <div className="file-summary"><div className="file-path">{path}</div></div>
> ```

Do you need nesting two times?

In any case, do the following (simplified; do not wrap a container element type around another container element):
```
<div>
   <span>{path}</span>
</div>
```

---

*[src/components/fileviewercov.js, line 11 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hDLqFLejxj6S0yC9L:-L-hDLqFLejxj6S0yC9M:b-pw11bz) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L11)):*
> ```JavaScript
>     this.state = {
>       expandTest: undefined,
>     };
> ```

No need to set this. Since it does not represent a data structure and you actually manipulate it anywhere in the class.

---

*[src/components/fileviewercov.js, line 56 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hH6g3xERi-zLnbI1Z:-L-hH6g4KJVXwzOT3Ebq:b-ky3egv) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L56)):*
> ```JavaScript
>       testTitle = `Line: ${lineNumber}`;
>       if (coverage.testsPerHitLine[lineNumber]) {
>         testList = this.getTestList(coverage.testsPerHitLine[lineNumber]);
> ```

I would have calculated all of this information at the top level component where the data is fetched.
I would have created a data structure like this (There might more data needed for each line):
```
this.state = {
   selectedLine: 0,
   lines: {
     0: {
       testList: getTestList(coverage.testsPerHitLine[lineNumber],
     }
   }
   allTests: getTestList(coverage.allTests),
}
```

That way all data processing happens in a single place and each piece in the top state has a direct influence on the UI.
Your UI is a representation of your state.

This also makes it very easy to access all significant data of a line anywhere in the app like this `this.state.lines[lineNumber].testList` or `this.state.allTests`.

I won't ask you to redo this.
I've filed it in here:
https://github.com/mozilla/firefox-code-coverage-frontend/issues/77

---

*[src/components/fileviewercov.js, line 66 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hEaaeqsSYZYOa4Udm:-L-hEaafWLnrecS0aDw6:bqrydww) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L66)):*
> ```JavaScript
>         <h3>{testTitle}</h3>
>         {testList}
>       </div>
> ```

Thank you. This looks cleaner.

---

*[src/components/fileviewercov.js, line 72 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hE44Gi5hIffIHyemP:-L-hE44HMoxVreIIkt8z:bzdle7a) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L72)):*
> ```JavaScript
> 
> // Test list item in the TestsSideViewer
> export const Test = ({ row, test, expand, handleTestOnExpand }) => (
> ```

Don't export it if you're not using it in other modules.

---

*[src/components/fileviewercov.js, line 76 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hEou0bNcX_1pkDKx8:-L-hEou0bNcX_1pkDKx9:bfwp3fb) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L76)):*
> ```JavaScript
>     <div className="toggleable-test-title" onClick={() => handleTestOnExpand(row)}>
>       <span className={`test-ptr ${expand}`}>&#x2023;</span>
>       <div className="test-name">
> ```

Make this a `span` unless I'm missing something (correctly if wrong).

---

*[src/components/fileviewercov.js, line 82 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hF2YJiBUSqCbDo0Cb:-L-hF2YJiBUSqCbDo0Cc:b56dmhj) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L82)):*
> ```JavaScript
>     <div className={`expandable-test-info ${expand}`}>
>       <ul className="test-detail-ul">
>         <li><span>platform : </span>{test.run.machine.platform}</li>
> ```

Why do you add `span` in here?

You can use a template literal (unless I'm mistaken):
```
<li>{`platform : ${test.run.machine.platform}`}</li>
```

---

*[src/components/fileviewercov.js, line 99 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hG_klC9eqLXwV8xD1:-L-hG_klC9eqLXwV8xD2:b-z2x0zo) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L99)):*
> ```JavaScript
>       <span className="coverage-percentage">
>         {((coveredLines / totalLines) * 100).toPrecision(4)}
>         % - {coveredLines} lines covered out of {totalLines} added
> ```

All of these coverage calculations I would have done at the top class when the coverage is fetched.
I won't ask you to rework it I just want you to be aware of this pattern I use.
Fetch data --> process it into all small state pieces that represent the UI.

---

*[src/utils/data.js, line 38 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hG8H-h_b5YZgRIKTi:-L-hG8H0kieARNtJ_bW-:bvlvjge) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/utils/data.js#L38)):*
> ```JavaScript
> };
> 
> /* get the coverage summary for a particular revision and file */
> ```

Use `//`

---

*[src/utils/data.js, line 39 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hJSy4T-McO5tNm1uh:-L-hJSy4T-McO5tNm1ui:b6u4i93) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/utils/data.js#L39)):*
> ```JavaScript
> 
> /* get the coverage summary for a particular revision and file */
> export const fileRevisionCoverageSummary = (coverage) => {
> ```

If you notice, some of the data manipulation you do at the `fileviewcov `module could have been done in here.

---

*[src/utils/data.js, line 170 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-h_kFICzlCePXV5Alg:-L-h_kFJFcZ6EwVSlNVv:b-uwrp9q) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/utils/data.js#L170)):*
> ```JavaScript
>       throw new Error();
>     }
>     return res.text();
> ```

Do the `split('\n')` here.

---

*[src/utils/fetch_data.js, line 12 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hMt_E80y3zxdMixEF:-L-hMt_FJb4QC2CGeN9r:b-3zdxvm) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/utils/fetch_data.js#L12)):*
> ```JavaScript
> };
> 
> const jsonPost = params =>
> ```

`({ url, body })` or make queryActiveData call it like ```jsonPost(`${activeData}/query`, body)```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-:-L-hNnCBm74Ao9LZZtC9:bb2spom)*
<!-- Sent from Reviewable.io -->

armenzg commented:
Filed an issue to follow up on another PR about the documentation issues I raised:
https://github.com/mozilla/firefox-code-coverage-frontend/issues/80

armenzg commented:
@LinkaiQi when do you estimate you will have this completed?
I'm only working until Thursday this coming week and then I'm off for 4 weeks.

LinkaiQi commented:
@yuenj will attempt to address current round of feedback. I think she will finish her last exam on Dec 21 (same as me, we are in the same university). There is a great chance that we are not able to complete is before Thursday.

yuenj commented:
Yes, my last exam is on the 20th.  @LinkaiQi does `3009478` address any of Armen's feedback?

armenzg commented:
OK. I've decided to look at the PR even when not officially working.
I will follow up with you via email.

yuenj commented:
Here is a link to test the changes: http://localhost:3000/#/file?revision=2d580aeac901&path=/accessible/atk/Platform.cpp

armenzg commented:
I will review the code in a bit, however, the most important thing to fix is this message:
"This branch cannot be rebased due to conflicts".

I assume you need to do this:

> git remote add upstream git@github.com:mozilla/firefox-code-coverage-frontend.git
> git pull upstream master

After you do that you will have to resolve all conflicts. If you're not comfortable at doing this you will have to do some tutorial and watch some videos. It is painful but it is part of getting the code ready to land.

armenzg commented:
I've deployed this branch to Heroku:
https://fccf-staging.herokuapp.com/#/file?revision=2d580aeac901&path=/accessible/atk/Platform.cpp

Before fixing *any* of my comments please make sure you fix the "branch cannot be rebased due to conflicts.". It's very important you focus on this first.

Also fix the following console error:

> Warning: Each child in an array or iterator should have a unique "key" prop.
> Check the render method of `FileViewer`. See https://fb.me/react-warning-keys
> for more information.

You need to look into fixing this overrun:
![image](https://user-images.githubusercontent.com/44410/34258833-cb4a8d6e-e62d-11e7-9cc9-0e3f7fbbd56d.png)

Another thing I've noticed is that this text speaks of "lines added" when we're not looking at a diff:
![image](https://user-images.githubusercontent.com/44410/34259301-c91a928a-e62f-11e7-8f8a-fb46765580c5.png)
Perhaps a better text would be "29 lines covered out of 134 coverable lines"

@klahnakoski f?
If I click on a white line I see "No test covers this line", however, when I click a red line I see all tests that cover this file.
Should white lines say "This is a line that is not coverable" and red lines say "There is no test that covers this line"? 

There's some weirdness in here (We can do a follow up; let's get this PR landed):
![image](https://user-images.githubusercontent.com/44410/34260840-2ef8ad26-e635-11e7-833e-a424149f8e4e.png)

armenzg commented:
Reviewed 3 of 7 files at r2, 2 of 7 files at r3, 1 of 2 files at r4.
Review status: 10 of 14 files reviewed at latest revision, 18 unresolved discussions.

---

*[docs/File_Coverage_View.pdf, line 0 at r4](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tOLvnYBgRx1FsdFVT:-L0tOLvo9CWMcbJAFOO3:b783qje) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/74f8d8ae119f7b836dc0a8e7cb780e057e20ff02/docs/File_Coverage_View.pdf#L0)):*
Please remove the .pdf and .jpeg files.
If you need these files for a follow up documentation PR remember to make a copy somewhere else.

---

*[src/style.css, line 180 at r4](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tOcVhBhWLsR0vBJpg:-L0tOcVibDkFYishAxYE:b-gto3i5) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/74f8d8ae119f7b836dc0a8e7cb780e057e20ff02/src/style.css#L180)):*
> ```CSS
> .file-meta-ul {
> 	float: right;
>   padding: 0;
> ```

Please fix the inconsistent indentation.

---

*[src/components/fileviewer.js, line 17 at r4](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tW-WJG1G0HXy0eDb0:-L0tW-WKudDaWT7VUlKy:bi85080) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/74f8d8ae119f7b836dc0a8e7cb780e057e20ff02/src/components/fileviewer.js#L17)):*
> ```JavaScript
>   }
> 
>   async componentDidMount() {
> ```

You might be surprised with this code but you can get rid of few "async" and "await" keywords:
```diff
armenzg@armenzg-mbp ccov_frontend$ git diff -U1
diff --git a/src/components/fileviewer.js b/src/components/fileviewer.js
index 285e03d..41764f6 100644
--- a/src/components/fileviewer.js
+++ b/src/components/fileviewer.js
@@ -16,5 +16,5 @@ export default class FileViewerContainer extends Component {

-  async componentDidMount() {
+  componentDidMount() {
     const { revision, path } = this.state;
-    await this.fetchData(revision, path, 'mozilla-central');
+    this.fetchData(revision, path, 'mozilla-central');
   }
@@ -30,3 +30,3 @@ export default class FileViewerContainer extends Component {

-  async fetchData(revision, path, repoPath = 'integration/mozilla-inbound') {
+  fetchData(revision, path, repoPath = 'integration/mozilla-inbound') {
     // Get source code from hg
@@ -42,3 +42,3 @@ export default class FileViewerContainer extends Component {
     try {
-      await Promise.all([fileSource(), coverageData()]);
+      Promise.all([fileSource(), coverageData()]);
     } catch (error) {
```

---

*[src/components/fileviewer.js, line 19 at r4](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tUXapygZmnH9R4lZI:-L0tUXapygZmnH9R4lZJ:b-2858wh) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/74f8d8ae119f7b836dc0a8e7cb780e057e20ff02/src/components/fileviewer.js#L19)):*
> ```JavaScript
>   async componentDidMount() {
>     const { revision, path } = this.state;
>     await this.fetchData(revision, path, 'mozilla-central');
> ```

When calling a function don't use a value that comes out of no where.
You can either create a constant at the top of the module (const REPOPATH = 'mozilla-central') OR have the function definition have a default value. In this case I prefer the later.

---

*[src/components/fileviewer.js, line 31 at r4](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tUvo6K8UQ2vnp9Rzo:-L0tUvo6K8UQ2vnp9Rzp:b-fthwuv) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/74f8d8ae119f7b836dc0a8e7cb780e057e20ff02/src/components/fileviewer.js#L31)):*
> ```JavaScript
>   }
> 
>   async fetchData(revision, path, repoPath = 'integration/mozilla-inbound') {
> ```

Always use 'mozilla-central' since all of our coverage data comes from that repository.

---

*[src/components/fileviewer.js, line 43 at r4](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tWH03oHX5zNhIP_Zq:-L0tWH03oHX5zNhIP_Zr:bx7u73q) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/74f8d8ae119f7b836dc0a8e7cb780e057e20ff02/src/components/fileviewer.js#L43)):*
> ```JavaScript
>     // Fetch source code and coverage in parallel
>     try {
>       await Promise.all([fileSource(), coverageData()]);
> ```

`Promise.all()` will not continue until all items in the array are resolved, thus, not needed to await for it.
If we remove `await` in here then there is no more usage of `await` in the body of `fetchData()`, thus, we can remove `async` from its definition.
I know we have `coverageData()` and `fileSource()` within it, however, they're self-contained within the `Promise.all()`

---

*[src/components/fileviewer.js, line 145 at r4](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tYFDraRyL-XQDxY2S:-L0tYFDraRyL-XQDxY2T:bt1yrzo) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/74f8d8ae119f7b836dc0a8e7cb780e057e20ff02/src/components/fileviewer.js#L145)):*
> ```JavaScript
>     } else {
>       msg = <span>&#x2714;</span>; // heavy checkmark
>     }
> ```

I would do this in one line. Not a blocker:
`const msg = (data) ? '&#x2714' : '&#x2026'; // Heavy checkmark of horizontal ellipsis`
Or even:
```javascript
const showStatus = (label, data) => (
   <li className="file-meta-li">
      {label}: {(data) ? '&#x2714' : '&#x2026'}  // Heavy checkmark of horizontal ellipsis
   </li>);
```

You can keep the span within the `<li>` even though I don't think there's need for one.

---

*[src/components/fileviewercov.js, line 11 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hDLqFLejxj6S0yC9L:-L0tgwq2ZzIaw7PDZuGi:b-flde9q) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L11)):*
<details><summary><i>Previously, armenzg (Armen Zambrano) wrote…</i></summary><blockquote>

No need to set this. Since it does not represent a data structure and you actually manipulate it anywhere in the class.
</blockquote></details>

I have not been able to figure out what this `expandTest` functionality is for. Would you mind removing it unless I'm missing something?

---

*[src/components/fileviewercov.js, line 75 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tmjOn7cXAUSWeANaQ:-L0tmjOoMjMNCTBCHYyD:b-hjlcoh) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L75)):*
> ```JavaScript
>   <li>
>     <div className="toggleable-test-title" onClick={() => handleTestOnExpand(row)}>
>       <span className={`test-ptr ${expand}`}>&#x2023;</span>
> ```

Could you please define this symbol as a constant at the top of the module and refer to it? ("magical string" issue)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76)*
<!-- Sent from Reviewable.io -->

armenzg commented:
Reviewed 3 of 7 files at r3, 1 of 2 files at r4.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[src/utils/data.js, line 170 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-h_kFICzlCePXV5Alg:-L0to8qanDBEQ34fjKrw:b-y3gl4) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/utils/data.js#L170)):*
<details><summary><i>Previously, armenzg (Armen Zambrano) wrote…</i></summary><blockquote>

Do the `split('\n')` here.
</blockquote></details>

Jennifer you can do the split in here if you do this:
```
return (await res.text()).split('\n');
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76)*
<!-- Sent from Reviewable.io -->

LinkaiQi commented:
> You need to look into fixing this overrun

How about adding a threshold width to display the TestsSideViewer? (solving the overrun issue: TestsSideViewer and FileViewer overlay each other)
 - If we don't have enough space in the right to show the TestsSideViewer (```browseWindowsSize``` - ```fileViewWidth``` < ```TestsSideViewerWidth```)
   displays message: "expand browser window to view test coverage"
- Else:
   displays TestsSideViewer normally

armenzg commented:
The overrun can easily be fixed using CSS grids.
https://mozilladevelopers.github.io/playground/css-grid
Consider the page as two columns. The sidebar being one of them.

Feel free to follow this up on a new PR after this gets landed.
I would rather get this PR landed.

Merry Christmas.

armenzg commented:
Reviewed 1 of 2 files at r5, 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/File_Coverage_View.pdf, line  at r4](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tOLvnYBgRx1FsdFVT:-L1wDMrnlVLE9Ed_8Ezp:bknu1ja) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/74f8d8ae119f7b836dc0a8e7cb780e057e20ff02/docs/File_Coverage_View.pdf#L)):*
<details><summary><i>Previously, armenzg (Armen Zambrano) wrote…</i></summary><blockquote>

Please remove the .pdf and .jpeg files.
If you need these files for a follow up documentation PR remember to make a copy somewhere else.
</blockquote></details>

Can you please remove the .jpeg and .pdf files?

---

*[src/components/fileviewercov.js, line 11 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L-hDLqFLejxj6S0yC9L:-L1wE28iF0DrtySedafw:ba1muv8) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L11)):*
<details><summary><i>Previously, armenzg (Armen Zambrano) wrote…</i></summary><blockquote>

I have not been able to figure out what this `expandTest` functionality is for. Would you mind removing it unless I'm missing something?
</blockquote></details>

Could you please address this?

---

*[src/components/fileviewercov.js, line 75 at r1](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76#-L0tmjOn7cXAUSWeANaQ:-L1wEJB9s1FwQSdMf_pA:beirj7w) ([raw file](https://github.com/mozilla/firefox-code-coverage-frontend/blob/0d479c7cd7d40e7cfaf1e48c6af2fc8389e70ddd/src/components/fileviewercov.js#L75)):*
<details><summary><i>Previously, armenzg (Armen Zambrano) wrote…</i></summary><blockquote>

Could you please define this symbol as a constant at the top of the module and refer to it? ("magical string" issue)
</blockquote></details>

I like how you solved putting magical strings into one module. Well done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/firefox-code-coverage-frontend/76)*
<!-- Sent from Reviewable.io -->

armenzg commented:
Good job to all!

I’ve decided to get the code landed as a single commit (except the docs/ directory):
https://github.com/mozilla/firefox-code-coverage-frontend/commit/671366674d84d1ea9ca8ed3248541e23ec1ff8f2

I would have preferred to let you deal with the merge conflicts, however, this will be easier.
Use the PR as a reference for all the work that went into this.

Can you please address my latest comments on follow up PRs (ignore the line about removing files from docs/)?:
https://github.com/mozilla/firefox-code-coverage-frontend/pull/76#issuecomment-355013934
